### ISSUE TITLE
Bad diagnostic when using `default_numeric_fallback`

### ISSUE THREAD

**Issue by c410-f3r** [rust-lang/rust-clippy#7130]
```rust
#![forbid(clippy::default_numeric_fallback)]

pub struct Foo<D> {
  pub data: D,
}

pub fn foo() -> Foo<i32> {
    Foo { data: 1 }
}
```

`foo()` clearly states that `data` is `i32` but `default_numeric_fallback` insists to explicitly declare `Foo { data: 1_i32 }`.

**Comment by camsteffen** [comment]
This looks like one of those unfortunate scenarios where we would have to reverse-engineer rustc's type inference logic.

**Comment by c410-f3r** [comment]
Closing due to the lack activeness. Feel free to open this issue if desired.

**Comment by camsteffen** [comment]
This seems like a good improvement to me. There are a lot of issues that have been inactive for a long time but are still valid.

**Comment by c410-f3r** [comment]
Closing due to inactivity (again). I am trying to clean up my list of opened issues so if desired feel to create another issue about the problem.
